File: solutions/Leetcode_1642/leetcode_1642.py
import heapq
from typing import List


class Solution1:
    def furthest_building(self, heights: List[int], bricks: int, ladders: int) -> int:
        h = []
        for i in range(1, len(heights)):
            if heights[i] - heights[i-1] > 0:
                heapq.heappush(h, heights[i] - heights[i-1])
                if len(h) > ladders:
                    if h[0] <= bricks:
                        bricks -= heapq.heappop(h)
                    else:
                        return i - 1
        return len(heights) - 1

File: solutions/Leetcode_1642/leetcode_1642.cc
#include <queue>
#include <vector>

using namespace std;

class Solution1
{
public:
    int furthestBuilding(vector<int> heights, int bricks, int ladders)
    {
        priority_queue<int> heap;
        for (int i = 0; i < heights.size() - 1; i++)
        {
            if (heights[i + 1] - heights[i] > 0)
                heap.push(heights[i] - heights[i + 1]);
            if (heap.size() > ladders)
            {
                bricks += heap.top();
                heap.pop();
            }
            if (bricks < 0)
                return i;
        }
        return heights.size() - 1;
    }
};
